src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/Transmission/OtlpExporterRetryTransmissionHandler.cs

```diff
@@ -0,0 +1,65 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+#nullable enable
+
+using Grpc.Core;
+using OpenTelemetry.Exporter.OpenTelemetryProtocol.Implementation.ExportClient;
+
+namespace OpenTelemetry.Exporter.OpenTelemetryProtocol.Implementation.Transmission;
+
+internal class OtlpExporterRetryTransmissionHandler<TRequest> : OtlpExporterTransmissionHandler<TRequest>
+{
+    internal OtlpExporterRetryTransmissionHandler(IExportClient<TRequest> exportClient)
+        : base(exportClient)
+    {
+    }
+
+    protected override bool OnSubmitRequestFailure(TRequest request, ExportClientResponse response)
+    {
+        var retryAttemptCount = 0;
+
+        while (retryAttemptCount < 5 && this.ShouldRetryRequest(request, response, retryAttemptCount++, out var sleepDuration))
```

Comment:
I think the way this is set up we will block the export thread from sending telemetry while we retry old failed stuff. Eventually the circular buffer will fill up and we'll start dropping new good stuff. Is that how we want it to work? I suppose we could have a background thread which will retry some number of failed requests it keeps around in memory? If we want that, I think a more derived `OtlpExporterInmemoryRetryTransmissionHandler` would be nice to house the in-memory logic.

Reply:
We do need to return ExportResult with each export request. If we process retries in a non-blocking way, we do need to think how we will handle ExportResult. Basically do we return ExportResult.Success or Failure?

Reply:
I gave my thoughts on this during our SIG meeting, but just capturing them here as well.

The [requirements of the OTLP exporter with regards to retry](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/exporter.md#retry) are centered around transient network errors:

> Transient errors MUST be handled with a retry strategy.

Transient errors are usually due to short-lived issues that self correct.

As it stands, the SDK's architecture is such that the export processor invokes its exporter synchronously. As @vishweshbankwar notes the contract between processor and exporter is a simple signal of success vs. failure.

For these reasons, I think we should keep things simple to begin with - at least for what is being termed "in memory retry". I propose the processor should block throughout any and all retry attempts not to exceed its configured timeout (which is 30 seconds by default). The concern that old stale data may cause new good data to get dropped is valid, but this concern already exists today. End users can adjust their batch processor's timeout to reduce this risk.

I have less to say regarding the requirements for persistent storage. This mechanism seems to suggest a need for handling failures that are beyond transient in nature - e.g., my backend was down for an hour. This kind of use case is not well represented by the specification today.